File: tracer.py
import json
import os


def trace(action: str, obj: dict) -> None:
    dest = os.getenv("AGENT_TRACE_FILE")
    if not dest:
        return
    with open(dest, "a") as fh:
        fh.write(json.dumps({"action": action, "details": obj}) + "\n")

File: src/tracer.go
package main

import (
	"encoding/json"
	"io"
)

// TraceEvent is one structured log line describing an orchestrator decision point.
type TraceEvent struct {
	Action  string         `json:"action"`
	Details map[string]any `json:"details,omitempty"`
}

// Tracer collects the events of one step execution.
type Tracer struct {
	Events []TraceEvent
}

func (t *Tracer) trace(action string, details map[string]any) {
	t.Events = append(t.Events, TraceEvent{Action: action, Details: details})
}

// writeEvents emits all collected events as JSON lines to w.
func writeEvents(w io.Writer, events []TraceEvent) {
	for _, e := range events {
		line, err := json.Marshal(e)
		if err != nil {
			continue
		}
		w.Write(line)
		w.Write([]byte("\n"))
	}
}
